lib/cache: if cached value supports it, run Shutdown on removal

miku opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

<!--
Describe the changes here
-->

Some backends implement the optional fs.Shutdowner interface. Until now,
Shutdown is only checked and called, when a backend is wrapped (e.g.
crypt, compress, ...).

To have a common way to perform operations at the end of the backend
lifecycle with proper error handling, we can call Shutdown at cache
clear time.

Since Shutdown returns an error, we add two new methods: cache.Err() and
cache.ClearErrors(); cache.Err() returns the error or errors that
occured. ClearErrors() resets any errors. All errors are recorded and
kept around until they are cleared.

The cache.Err() approach is inspired from similar implementations, e.g.
in bufio.Scanner. This way we can leave method signatures in lib/cache
unchanged and we also do not emphasize the error reporting (but allow
for checks).

#### Was the change discussed in an issue or in the forum before?

<!--
Link issues and relevant forum posts here.
-->

Yes, previous discussion: https://forum.rclone.org/t/31336

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented on lib/cache/cache.go:
```diff
@@ -16,6 +20,7 @@ type Cache struct {
 	expireRunning  bool
 	expireDuration time.Duration // expire the cache entry when it is older than this
 	expireInterval time.Duration // interval to run the cache expire
+	err            error         // any error, e.g. while dropping backends which can shutdown
```
Instead of storing the error here, I'd suggest adding a value

```go
finalize func(value interface{})
```

and a method

```go
func (c *Cache) SetFinalizer(finalize func(value interface{}))
```

Which can be used to have a callback for finalizing objects.

That will look a lot like `tryShutdown` except it will use `fs.Errorf` to log each error, and it will call `fs.CountError` to set the global error flag.

func (c *Cache) AddFinalizer(func()

ncw commented on lib/cache/cache.go:
```diff
@@ -154,7 +159,12 @@ func (c *Cache) GetMaybe(key string) (value interface{}, found bool) {
-	_, found := c.cache[key]
+	entry, found := c.cache[key]
+	if found {
+		if err := tryShutdown(entry.value); err != nil {
```
This then becomes

```go
if c.finalize != nil {
    c.finalize(entry.value)
}
```

I'd probably make sure c.finalize is never nil so points to a no-op function so you can do without the `if c.finalize { }` which will make the code a lot neater.

ncw reviewed (commented):
I think the basic idea looks sound :-)

I've put some suggestions for rework which makes the `lib/cache` free from `fs` and makes the error handling a bit easier.

Thank you :-)

miku commented:
Thanks a lot for your review, I'm just getting into the rclone codebase; the finalizer approach is much nicer. I addressed your feedback and hope the code looks a bit better now.

ncw commented on cmd/cmd.go:
```diff
@@ -321,6 +321,14 @@ func Run(Retry bool, showStats bool, cmd *cobra.Command, f func() error) {
 		}
 	}
 
+	// add hook to wind down backends
```
I think this should go here

https://github.com/rclone/rclone/blob/326c43ab3f5c7c8570bf87b4656b41846ce00877/fs/cache/cache.go#L22-L27

ncw commented on fs/cache/cache.go:
```diff
@@ -167,3 +167,9 @@ func Entries() int {
 	createOnFirstUse()
 	return c.Entries()
 }
+
```
... then we don't need this method

ncw reviewed (commented):
That is looking really good now :-)

I think we should move the call to SetFinalizer and I've put a couple of notes inline about that.

Thank you :-)

miku commented:
Thank you! - This will greatly improve error handling options for our backend implementation (still in development).

ncw commented:
And thank you for doing this useful, but unglamorous bit of work :-)
